### PR TITLE
fixes #32518 - Adding Ansible fact parser

### PR DESCRIPTION
**Pull request by domitea**
Adding Ansible fact parser for better refactoring of parsers logic

### CODE REVIEW

File: Gemfile
Comment (tbrisker) [inline Gemfile]: I think these belong in `bundler.d/Gemfile.local.rb` 😉

File: app/services/ansible_fact_parser.rb
Comment (ekohl) [inline app/services/ansible_fact_parser.rb:124]:
```suggestion
    fact_names.each do |fact_name|
      value = facts[fact_name]
      return value if value.present?
    end
```

File: app/services/operating_system_parser.rb
Comment (ekohl) [inline app/services/operating_system_parser.rb]: This feels like a bad design. It's strictly tied to Ansible, but nothing in the name of the class suggests that. If anything, It should be named `AnsibleOperatingSystemParser`.

File: test/unit/ansible_fact_parser_test.rb
Comment (ekohl) [inline test/unit/ansible_fact_parser_test.rb]: I'm more of a fan of using `let`and `subject`:
```suggestion
    let(:facts_json) { HashWithIndifferentAccess.new(read_json_fixture('facts/ansible_facts.json')) }
    subject { AnsibleFactParser.new(facts_json) }
```

Then in the actual test you use `subject.domain`. Maybe it's because I'm used to rspec. @tbrisker what do you think?

File: app/services/operating_system_parser.rb
Comment (ares) [reply app/services/operating_system_parser.rb]: yeah, originally this code lived in `ForemanAnsible` namespace so it made sense, but as part of the move, the class should be either renamed or kept under `ForemanAnsible` module. I don't mind much which way, since after this is in, we should start deduplicating the logic and my hope is, as part of that effort, this class will be dropped entirely.

File: test/unit/ansible_fact_parser_test.rb
Comment (ares) [reply test/unit/ansible_fact_parser_test.rb]: I'm not the person you asked, but :-)

```
grep -r setup test | wc -l
1064

grep -r let test | wc -l  
1616
```

while it's not entirely fair comparison since `let` is used per variable, it's safe to say we don't have a single way in our code and this is a personal preference. While I also prefer `let`, I think we, as maintainers, shouldn't be asking for this and also not in the PR that moves the existing code.

Comment (ekohl) [reply test/unit/ansible_fact_parser_test.rb]: Perhaps it would be good to mention in the commit message that this moves over files from an existing repository and isn't really new code.

Comment (ares) [reply test/unit/ansible_fact_parser_test.rb]: I agree, also @domitea the commit shouldn't say Fixes but Refs, since the redmine issue currently tracks all other parsers too. Or create a separate RM issue per parser and use that number.

Comment (domitea) [reply test/unit/ansible_fact_parser_test.rb]: I've updated the commit message.

**Review by ekohl — COMMENTED** [review]
I'm ok with merging this once CI is green.

Some related work to make review easier: https://github.com/theforeman/foreman_ansible/pull/408 drops the plugin code.

**Review by ares — DISMISSED** [review]
This does not work without additional classes that are required. Parser only parses the data to more generic form, but we also need the Importer, otherwise facts are not stored in the DB. That also implies we need to move `ForemanAnsible::FactName` to core. And with that, we'll also need the origin icon. Since we'll be moving at least some class with the ForemanAnsible namespace, I'd prefer to change `AnsibleOperatingSystemParser` back to `ForemanAnsible::OperatingSystemParser`

https://github.com/theforeman/foreman_ansible/blob/master/app/services/foreman_ansible/fact_importer.rb
https://github.com/theforeman/foreman_ansible/blob/7876787e5e5cd9a08f4d86b1327d1bec558a0615/app/models/foreman_ansible/fact_name.rb#L7
https://github.com/theforeman/foreman_ansible/blob/master/app/assets/images/foreman_ansible/Ansible.png

**Comment by ekohl** [comment]
> Since we'll be moving at least some class with the ForemanAnsible namespace, I'd prefer to change `AnsibleOperatingSystemParser` back to `ForemanAnsible::OperatingSystemParser`

I'm ok with importing the namespace (to show its history) and then refactor it into a proper core parser.

In https://github.com/theforeman/foreman/pull/8451 I made a start with this. I'd be happy if you took over and really unified the fact parsers.

Also, a common OS parser is probably possible once you have a really common fact parser that provides the correct abstractions. Looking forward to seeing that happen.

**Comment by domitea** [comment]
> In #8451 I made a start with this. I'd be happy if you took over and really unified the fact parsers.
> 
> Also, a common OS parser is probably possible once you have a really common fact parser that provides the correct abstractions. Looking forward to seeing that happen.

Yes, I have this in mind too and I would like to unify the parsers and their tests. 
However I'm not sure if it'll be a common OS parser in future, right know I just took them together with the parser because it's dependent on them and I don't want to do any futher refactoring. Same situation is a `fact_sparser` that is used in importer.

File: test/unit/foreman_ansible/fact_parser_test.rb
Comment (lzap) [inline test/unit/foreman_ansible/fact_parser_test.rb:154]: I suggest to pull this out into fixtures/ directory as a separate file, we will be adding more and more over time (multiple versions, OSes etc).

**Comment by lzap** [comment]
For the record tests are passing, exept the commit message.

**Comment by domitea** [comment]
@lzap I would like to do any refactoring after merging.

**Review by lzap — DISMISSED** [review]
What are we waiting for then? Let's merge and refactor later? Any objections?

**Comment by ares** [comment]
> What are we waiting for then? Let's merge and refactor later? Any objections?

I haven't tested, but no objections from my side. The sooner it gets in, the sooner we can further improve it :-)

**Comment by lzap** [comment]
Can you rebase, CI had a temporary problem with git, let's get all those PRs merged so you can start working on the refactoring. Please add more fixtures during refactoring.

**Comment by lzap** [comment]
The moment this (and any other similar parsing PR) is merged we need to bump required minimum foreman version in each plugin to 2.6 and remove the parser from the plugin.

File: config/initializers/z_plugin_parsers.rb
Comment (lzap) [inline config/initializers/z_plugin_parsers.rb]: So this is intentionally moved to the end of initializer chain in Rails so every existing fact parser by a plugin is overridden with this.

It could be useful to change `register_fact_parser` method to issue a warning into logger if there already was an existing parser registered to a particular name and which class was that. So we know if this happens for some users who would e.g. forget to upgrade their plugins for any reason.

File: app/models/fact_names/foreman_ansible/fact_name.rb
Comment (lzap) [inline app/models/fact_names/foreman_ansible/fact_name.rb:14]: Can you create a TODO item (Redmine perhaps): Move images from app/assets/images/xyz.png into icons16x16 and icons64x64 subdirectories? We now have two sizes of images which allows us to show big images. We do not do this yet, but we can in the future.

File: app/services/foreman_ansible/operating_system_parser.rb
Comment (lzap) [inline app/services/foreman_ansible/operating_system_parser.rb:40]: This could be moved into operatingsystems/debian.rb at some point, do not do that now but make a note.

**Review by lzap — APPROVED** [review]
It works fine, it looks like when there is an array the parser does not correctly parse but this is I think a bug we had for quite some time. We should fix this at some point, for those cases there must be a key from a subhash picked as the main key instead of an array:

image

**Comment by lzap** [comment]
I am requesting re-review from all the others. This is good to go.

**Review by tbrisker — APPROVED** [review]
Code wise LGTM, lets get this and the other parsers in for 3.0 and then refactor to improve.

**Review by ekohl — APPROVED** [review]
Same approval as https://github.com/theforeman/foreman/pull/8491#pullrequestreview-651276512 (which was dismissed) and I forgot to approve again. This needs an update in a follow up PR, but everyone's aware of that. First this needs to be merged.

**Pull request merged**
